### ISSUE TITLE
Support for zenodo caching  of scheduled cluster jobs?

### ISSUE THREAD

**Issue by palumbom** [showyourwork/showyourwork#180]
(@rodluger is aware of most of the details of this use case issue, but here is a concise-ish summary)

I run simulations which are too computationally expensive for a laptop/GitHub Actions, etc. So for things to work nicely with showyourwork, I need to generate my output files on a cluster and then cache them with Zenodo, as described in the [documentation](https://show-your.work/en/latest/zenodo/). 

But because I need to submit jobs with a batch scheduler, I cannot invoke ```showyourwork build``` to run my simulations on the cluster. Instead, I generate my output files outside of ```showyourwork build``` via the job scheduler. After my jobs have run, I am able to successfully build the paper locally (after copying over the output from the cluster) or on the cluster, but showyourwork does not cache the files -- [see this log](https://github.com/showyourwork/showyourwork/files/9530391/showyourwork.log) generated from this [project](https://github.com/palumbom/palumbo22) -- so GitHub Actions and anyone who clones the repo won't be able to generate the PDF. 

It sounds like the above works as intended, since you shouldn't be able to "trick" showyourwork into using any old output file and circumvent true reproducibility. But as currently implemented, I simply can't run ```showyourwork build``` on the cluster and need to rely on a hack privately communicated to me by @rodluger. 

It would be awesome to have caching play nicely with scheduled runs. I'm not sure I have the expertise necessary to help with the implementation of this, but I am happy to be a guinea pig if it contributes to the development :) 

TL;DR -- could showyourwork implement a way to reconcile scheduled cluster jobs with Zenodo caching?

**Comment by rodluger** [comment]
In principle, since `showyourwork` is basically just a (complicated) wrapper around `Snakemake`, we should be able to take advantage of its [cluster execution features](https://snakemake.readthedocs.io/en/stable/executing/cluster.html). All unused arguments to `showyourwork build` get forwarded to `snakemake`, so instead of (e.g.)

```
snakemake --cluster qsub --jobs 32
```

you could try

```
showyourwork build --cluster qsub --jobs 32
```

I haven't tested this, however, and I can imagine a few reasons why this might fail. @palumbom I'd be curious to see if you can get anywhere with this approach.

If that doesn't work out of the box, I think we can definitely bootstrap something on our end, since the functionality is definitely there in `Snakemake`!

**Comment by palumbom** [comment]
Oh sweet, I didn't know ```snakemake``` had those features. I'm a bit tied up this week, but I'll definitely tinker around and report back ~next week with what I find!

**Comment by palumbom** [comment]
I've got a little bit of an update. 

I was able to get ```snakemake``` to  submit jobs to the cluster by using the [profiles feature](https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles). I used [this template profile](https://github.com/jdblischak/smk-simple-slurm), and filled-in template is [here](https://github.com/palumbom/palumbo22/tree/main/profile/simple). I can override the default resources in the profile by requesting different resources in the Snakefile rules (see this [example](https://github.com/palumbom/palumbo22/blob/main/Snakefile)). 

I can tell ```snakemake``` to use this profile to submit jobs to the cluster by invoking:

```bash
snakemake -s  Snakefile --profile ./profile/simple
```

```showyourwork``` is happy to pass along the ```--profile``` argument just fine when I run:

```bash
showyourwork build --profile ./profile/simple
```

However, if I include the ```snakemake``` ```--cluster-cancel``` option in my profile, I get told that ```snakemake``` does not recognize the ```--cluster-cancel``` argument. See the attached [output.txt](https://github.com/showyourwork/showyourwork/files/9709598/output.txt). 

To actually test the full build process, I'm currently running ```showyourwork build --profile ./profile/simple``` in it's own scheduled job (see [build.sh](https://github.com/palumbom/palumbo22/blob/main/build.sh)). Inside this job, ```showyourwork``` will then request additional scheduled jobs to produce my figures. As I understand it, this is done sequentially. So ```showyourwork``` will use ```snakemake``` to submit a job to produce Figure 1, wait for that job to run, then submit a job to produce Figure 2, wait for that job to run, and so on.

I'm out of time to investigate today, but do you know if there a way to get this to happen in parallel? I.e., the main job submits jobs for Figures 1-N simultaneously, and then once all of them have run, it actually goes and builds the paper? 

For now it seems like things are running fine, but I won't know for sure for a few days, since Figures 4-7 are all overnight jobs.

**Comment by rodluger** [comment]
> However, if I include the snakemake --cluster-cancel option in my profile, I get told that snakemake does not recognize the --cluster-cancel argument. See the attached [output.txt](https://github.com/showyourwork/showyourwork/files/9709598/output.txt).

The current version of `showyourwork` pins `Snakemake==6.15.5`. The `cluster-cancel` feature was [introduced](https://snakemake.readthedocs.io/en/stable/project_info/history.html#id101) in the next release, `7.0.0`. I'm actually working on upgrading `Snakemake` (#212), so the next version of `showyourwork` should support this feature.

> I'm out of time to investigate today, but do you know if there a way to get this to happen in parallel? I.e., the main job submits jobs for Figures 1-N simultaneously, and then once all of them have run, it actually goes and builds the paper?

Have you tried specifying the number cores explicitly? `showyourwork` passes `-c1` (one core) by default, but you can override it, e.g., 

```
showyourwork build --cores 10
```

See [here](https://snakemake.readthedocs.io/en/stable/tutorial/advanced.html#step-1-specifying-the-number-of-used-threads).

**Comment by palumbom** [comment]
> The current version of ```showyourwork``` pins ```Snakemake==6.15.5```. The ```cluster-cancel``` feature was [introduced](https://snakemake.readthedocs.io/en/stable/project_info/history.html#id101) in the next release, ```7.0.0```. I'm actually working on upgrading ```Snakemake``` (https://github.com/showyourwork/showyourwork/pull/212), so the next version of ```showyourwork``` should support this feature.

Good to hear! ```cluster-cancel``` isn't a make-or-break, but it is convenient to have :)

> Have you tried specifying the number cores explicitly? ```showyourwork``` passes ```-c1``` (one core) by default, but you can override it, e.g.,

> ```showyourwork build --cores 10```
> See [here](https://snakemake.readthedocs.io/en/stable/tutorial/advanced.html#step-1-specifying-the-number-of-used-threads).

Ah interesting. I *think* this would work, but would be a suboptimal solution. Since I need to run ```showyourwork build``` in its own job, I would need to request N cores for the main job, which would use those N cores to submit and monitor N jobs, each of which requires M cores. So each of those N cores would spend all their time waiting for the job it is monitoring to finish, which seems wasteful. 

That being said, I understand that this is a fairly niche problem, and that ultimately this may be limited by the capabilities of ```snakemake``` and the idiosyncrasies of my computing setup. If I end up having to request a couple semi-superfluous cores, I will live :)

**Comment by rodluger** [comment]
Hmm. Why do you need to run `showyourwork` from within its own cluster job? Shouldn't you be able to run it from the login node?

**Comment by palumbom** [comment]
> Why do you need to run showyourwork from within its own cluster job? Shouldn't you be able to run it from the login node?

As far as I understand, because of some idiosyncrasies of the system. Nominally, we are *only* supposed to use login nodes to submit jobs. All other actual computing is supposed to be done via interactive jobs or queued jobs. 

Practically, the two-factor authentication boots me off of ssh connections to submit nodes when left overnight, even with a ```ServerAliveInterval``` set in my ```.ssh/config```. Last I tried, using ```screen``` as a workaround to this problem fails too, since I think anything seen as a stray process is automatically killed (and the definition of "stray" seems pretty generous). 

I also just tried passing ```--cores N``` to ```showyourworkbuild```, and it looks like it is not taking advantage of those cores. I'm guessing this is because I haven't set up the ```snakemake``` rules and cluster config file quite properly to deal with this, but I'll have to check into that for sure another day.

In the meantime, it appears that things are building and they will hopefully cache correctly. I'll let you know in a couple days once everything has run.

**Comment by HealthyPear** [comment]
@palumbom this issue is very old and the project advanced quite a bit in the meanwhile

now we support snakemake 9!

if you still use showyourwork (or want to resume using it) would you be able to make a update about what is still open about this issue?

you (or anybody else stumbling here) might also want to take a look at #297 in case you have a proposition for this kind of use-case!

pinging also our zenodo expert @vandalt 😉  in case he has already experience with zenodo chaching +clusters

**Comment by vandalt** [comment]
I don't have experience with Zenodo caching + clusters together, but I have experience with both separately, and would really like to use them together soon-ish.

I'll see if I can run a simple `showyourwork` workflow on the cluster and test it with Zenodo as a first step.
